### ISSUE TITLE
[Haskell] Some errors/warnings are detected but do not register with spacemacs as syntax errors

### ISSUE THREAD

**Issue by tekul** [syl20bnr/spacemacs#3052]
I've noticed that some errors and warnings in Haskell files are shown in the gutter as either '?' (warnings) or '!' errors, but are not shown in the totals for the file and are not navigable using SPC-e-n etc. Also you can't actually find out what the error text is in these cases.

For example, here is a snapshot of a buffer with two warnings in view, one of which is for the  the `cc` variable not being used and the other for use of the `maybe` function instead of `fromMaybe`:

![untitled](https://cloud.githubusercontent.com/assets/191720/9913184/76491922-5ca3-11e5-837f-a15ba12f4b46.png)

The latter shows up nicely in spacemacs, and will display the error text for you, but the former only shows as a '?' in the gutter, so it isn't possible to tell if a file is error free without either checking all the gutter or actually trying to compile the file. 

The first warning is from ghc-mod and the second from hlint, but that doesn't seem to be the issue - I see ghc-mod warnings and errors fine in other files without any problem. I'm using ghc-mod 5.4, but it's an issue I've seen with earlier versions too. Obviously something is detecting the warning, and indeed, if I run from the command line I do see it:

```
$ ghc-mod check Crypto/PubKey/RSA/OAEP.hs
...
Crypto/PubKey/RSA/OAEP.hs:32:1:Warning: The import of ‘ByteArrayAccess’from module ‘Crypto.Internal.ByteArray’ is redundant
Crypto/PubKey/RSA/OAEP.hs:69:11:Warning: Defined but not used: ‘cc’
```

Just wondering if someone has seen this and knows what's causing it. I can't quite wrap my head round how all the different syntax checkers work together.

**Comment by TheBB** [comment]
I suspect the ? and ! are from flymake, which is an older tool distributed with emacs. Spacemacs uses flycheck, which displays errors as circles and  has its place in the modeline. Check if flymake has been incorrectly enabled.

**Comment by tekul** [comment]
Thanks, but I haven't done anything flycheck related - I just enable haskell mode in my spacemacs file. How would I correct whether it is incorrectly enabled (there are no startup errors reported)?

Also, the missing error is detected by ghc-mod (which I believe spacemacs also uses), which isn't part of flycheck. The second error (which does display correctly) is presumably reported by flycheck's hlint checker, so flycheck is working OK there.

**Comment by syl20bnr** [comment]
I guess a recent update of ghc-mod added fringe bitmaps. The underline of it are masked when flycheck is active: https://github.com/syl20bnr/spacemacs/blob/develop/layers/+lang/haskell/packages.el#L48-L51

**Comment by tekul** [comment]
Hi Sylvain. Thanks for that. I'm still a bit confused though. Does that mean that no ghc-mod errors are supposed to be shown if flycheck is enabled? If so, why is ghc-mod still being used to check the file? And wouldn't flycheck be expected to detect the same issues?

Also, I see similar warnings in other files which do display properly.

**Comment by syl20bnr** [comment]
I guess that flycheck should display them. It is like flycheck checker using ghc is not enabled.

**Comment by tekul** [comment]
OK thanks. I'll see if I can work out what commands the flycheck checker is running and try them myself.

**Comment by tekul** [comment]
Ok, I've worked out what the issue was. The flycheck ghc checker was failing silently on an imported file because of this issue related to CPP-style macros:

https://github.com/flycheck/flycheck-haskell/issues/39

Following the advice in there I was able to get it to work by creating a separate macros file and setting

```
 flycheck-ghc-args '("-DWITH_DEEPSEQ_SUPPORT" "-optP-include" "-optP/Users/luke/Haskell/default_macros.h")
```

I can then see the question mark appearing and then being overwritten by a normal warning marker, and I get the usual information I expect in the editor.

So it isn't something spacemacs really needs to worry about. It still seems unnecessary to run both checkers if the ghc-mod output is being ignored though (unless it's needed for some other information).

Also, I noticed that the [Haskell readme](https://github.com/syl20bnr/spacemacs/blob/master/contrib/!lang/haskell/README.org#features) says that it "Provides syntax highlighting for C-source", which doesn't sound right :-).

Thanks a lot for the pointers which helped me to solve this issue.
